src/crypto/secp156k1_provider/secp256k1_provider_impl.cpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "libp2p/crypto/secp256k1_provider/secp256k1_provider_impl.hpp"
+
+#include <memory>
+
+#include <openssl/bn.h>
+#include <openssl/ec.h>
+#include <openssl/obj_mac.h>
+
+#include "libp2p/crypto/error.hpp"
+#include "libp2p/crypto/sha/sha256.hpp"
+
+namespace libp2p::crypto::secp256k1 {
+  outcome::result<KeyPair> Secp256k1ProviderImpl::generateKeyPair() const {
+    PublicKey publicKey{};
+    PrivateKey privateKey{};
+    std::shared_ptr<EC_KEY> key{EC_KEY_new_by_curve_name(NID_secp256k1),
+                                EC_KEY_free};
+    if (EC_KEY_generate_key(key.get()) != 1) {
+      return KeyGeneratorError::KEY_GENERATION_FAILED;
+    }
+    const BIGNUM *privateNum = EC_KEY_get0_private_key(key.get());
+    if (BN_bn2binpad(privateNum, privateKey.data(), privateKey.size()) < 0) {
+      return KeyGeneratorError::KEY_GENERATION_FAILED;
+    }
+    EC_KEY_set_conv_form(key.get(), POINT_CONVERSION_COMPRESSED);
+    auto publicKeyLength = i2o_ECPublicKey(key.get(), nullptr);
+    if (publicKeyLength != publicKey.size()) {
+      return KeyGeneratorError::KEY_GENERATION_FAILED;
+    }
+    uint8_t *publicKeyPtr = publicKey.data();
+    if (i2o_ECPublicKey(key.get(), &publicKeyPtr) != publicKeyLength) {
+      return KeyGeneratorError::KEY_GENERATION_FAILED;
+    }
+    return KeyPair{privateKey, publicKey};
+  }
+
+  outcome::result<PublicKey> Secp256k1ProviderImpl::derivePublicKey(
+      const PrivateKey &privateKey) const {
+    auto key = bytesToPrivateKey(privateKey);
+    if (key.has_error()) {
+      return key.error();
+    }
+    PublicKey publicKey{};
+    uint8_t *publicKeyPtr = publicKey.data();
+    ssize_t generatedLength = i2o_ECPublicKey(key.value().get(), &publicKeyPtr);
+    if (generatedLength != publicKey.size()) {
+      return KeyGeneratorError::KEY_GENERATION_FAILED;
+    }
+    return publicKey;
+  }
+
+  outcome::result<Signature> Secp256k1ProviderImpl::sign(
+      gsl::span<uint8_t> message, const PrivateKey &privateKey) const {
+    common::Hash256 digest = sha256(message);
+    auto key = bytesToPrivateKey(privateKey);
+    if (key.has_error()) {
+      return key.error();
+    }
+    std::shared_ptr<ECDSA_SIG> ecSignature{
+        ECDSA_do_sign(digest.data(), digest.size(), key.value().get()),
+        ECDSA_SIG_free};
+    if (ecSignature == nullptr) {
+      return CryptoProviderError::SIGNATURE_GENERATION_FAILED;
+    }
+    Signature signature;
+    ssize_t signatureLength = i2d_ECDSA_SIG(ecSignature.get(), nullptr);
+    if (signatureLength < 0) {
+      return CryptoProviderError::SIGNATURE_GENERATION_FAILED;
+    }
+    signature.resize(signatureLength);
+    uint8_t *signaturePtr = signature.data();
+    i2d_ECDSA_SIG(ecSignature.get(), &signaturePtr);
+    return signature;
+  }
+
+  outcome::result<bool> Secp256k1ProviderImpl::verify(
```

Comment:
What about `outcome::result<void>` here? Won't it be less redundant?

Reply:
We have 3 cases:

1. Signature verifying failed
2. Verifying successful, signature invalid
3. Verifying successful, signature valid